source/extensions/quic_listeners/quiche/platform/quic_mem_slice_impl.h

```diff
@@ -0,0 +1,85 @@
+#pragma once
+
+// NOLINT(namespace-envoy)
+
+// This file is part of the QUICHE platform implementation, and is not to be
+// consumed or referenced directly by other Envoy code. It serves purely as a
+// porting layer for QUICHE.
+
+#include <cstddef>
+#include <iostream>
+#include <memory>
+
+#include "common/buffer/buffer_impl.h"
+
+#include "quiche/quic/core/quic_buffer_allocator.h"
+
+namespace quic {
+
+// Implements the interface required by
+// https://quiche.googlesource.com/quiche/+/refs/heads/master/quic/platform/api/quic_mem_slice.h
+class QuicMemSliceImpl {
+public:
+  // Constructs an empty QuicMemSliceImpl.
+  QuicMemSliceImpl() = default;
+
+  // Constructs a buffer fragment and the buffer of given length the fragment points to with the
+  // guarantee that buffer and fragment are both aligned according to max_align_t.
+  static Envoy::Buffer::BufferFragmentImpl*
+  allocateBufferAndFragment(QuicBufferAllocator* allocator, size_t length);
+
+  // Constructs a QuicMemSliceImpl by let |allocator| allocate a data buffer of
+  // |length|.
+  QuicMemSliceImpl(QuicBufferAllocator* allocator, size_t length);
+
+  // Constructs a QuicMemSliceImpl from a Buffer::Instance with first |length| bytes in it.
+  // Data will be moved from |buffer| to this mem slice.
+  // Prerequisite: |buffer| has at least |length| bytes of data and not empty.
+  explicit QuicMemSliceImpl(Envoy::Buffer::Instance& buffer, size_t length);
+
+  QuicMemSliceImpl(const QuicMemSliceImpl& other) = delete;
+  // Move constructors. |other| will not hold a reference to the data buffer
+  // after this call completes.
+  QuicMemSliceImpl(QuicMemSliceImpl&& other) noexcept { *this = std::move(other); }
+
+  QuicMemSliceImpl& operator=(const QuicMemSliceImpl& other) = delete;
+  QuicMemSliceImpl& operator=(QuicMemSliceImpl&& other) noexcept {
+    if (this != &other) {
+      single_slice_buffer_.move(other.single_slice_buffer_);
+    }
+    return *this;
+  }
+
+  // Below methods implements interface needed by QuicMemSlice.
+  void Reset() { single_slice_buffer_.drain(length()); }
+
+  // Returns a char pointer to the one and only slice in buffer.
+  const char* data() const;
+
+  size_t length() const { return single_slice_buffer_.length(); }
+  bool empty() const { return length() == 0; }
+
+private:
+  // Used to align both fragment and buffer at max aligned address.
+  struct BufferFragmentBundle {
+    static BufferFragmentBundle* CreateBundleWithSize(QuicBufferAllocator* allocator,
+                                                      size_t length) {
+      return reinterpret_cast<BufferFragmentBundle*>(
+          allocator->New(sizeof(BufferFragmentBundle) + length));
+    }
+
+    // TODO(danzh) fragment_ is not aligned in memory. This can cause extra
```

Comment:
The only reason fragment_ would be unaligned is if QuicBufferAllocator* returns an unaligned address. If you were just to replace CreateBundleWithSize() with a simple call to `malloc(sizeof(BufferFragmentBundle) + length))` it would be aligned.

It's not clear from 
https://quiche.googlesource.com/quiche/+/ff86db3f4a3d59250eeafbb9787fb52b70e6ac31/quic/core/quic_buffer_allocator.h?autodive=0%2F%2F that QuicBufferAllocator is appropriate for this. It returns a char* which AFAICT may be unaligned (e.g. it could be allocating from any offset from a larger char-buf pool). in which case it's a crasher and not just an optimization to fail to align it.

But IMO that should be done by calling an API that's explicitly guaranteed to return an aligned result. Any reason not to just call new/delete or malloc/free here?

Reply:
> The only reason fragment_ would be unaligned is if QuicBufferAllocator* returns an unaligned address. If you were just to replace CreateBundleWithSize() with a simple call to `malloc(sizeof(BufferFragmentBundle) + length))` it would be aligned.

Currently QuicBufferAllocator just uses new char[]. But this still should return aligned memory AFAIK. Otherwise how would `new char[sizeof(T)]` work in general? The returned address might not be 8 byte aligned, but this should just be an optimization issue rather than crash point.

And I believe the implementation of new() wraps malloc().

> It's not clear from
> https://quiche.googlesource.com/quiche/+/ff86db3f4a3d59250eeafbb9787fb52b70e6ac31/quic/core/quic_buffer_allocator.h?autodive=0%2F%2F that QuicBufferAllocator is appropriate for this. It returns a char* which AFAICT may be unaligned (e.g. it could be allocating from any offset from a larger char-buf pool). in which case it's a crasher and not just an optimization to fail to align it.

new[] 
> But IMO that should be done by calling an API that's explicitly guaranteed to return an aligned result. Any reason not to just call new/delete or malloc/free here?

QuicBufferAllocator is the API QUIC uses to have better memory management. In some platform, we use buffer pool underneath that interface. Envoy quic may need that one day though its implementation is just calling :new for now.

Reply:
Why don't we use use stick with new[] then? Envoy is already customizing its memory allocator by linking in tcmalloc. Why do we want to also have another memory allocator added here?

Reply:
> Why don't we use use stick with new[] then? 

Sorry, I don't understand this.

> Envoy is already customizing its memory allocator by linking in tcmalloc. Why do we want to also have another memory allocator added here?

In the future, QuicBufferAllocator might do something different from new[], i.e. allocate memory from Quic's own buffer pool. Using an interface here instead of direct new[] allows us to customize memory allocation for quic specifically.

Reply:
Then I think this is speculative generality, no? We can always switch to a custom allocator when we need one. In the meantime remove the indirection and just new/delete straight up?

But anyway I have another idea to clean up this code. I took a pattern from @brian-pane and tried to generalize it in https://gist.github.com/jmarantz/bfc3395dfc8b8e0c8c64ffa2af96b53e -- this example just makes an inlined string allocator, holding the size and the bytes in contiguous memory. But it factors out the operator new/delete override magic to a helper class that could be shared with Buffer::OwnedSlice, and we can switch some other patterns in the codebase to that one.

What do you think of switching to that pattern?

Could be a follow-up.

Reply:
Why would Quic need its own buffer allocator? How would Quic even know which allocator was used for this? Isn't the connection with the deletion-mechanism captured here in the API?

Reply:
> Why would Quic need its own buffer allocator? 
In case that Quic needs to copy over data instead of move the underlying memory. And this open a door for memory usage optimization such as buffer pool.

> How would Quic even know which allocator was used for this? Isn't the connection with the deletion-mechanism captured here in the API?
It's chose by selecting a QuicBufferAllocator subclass in QuicStreamBufferAllocatorImpl. The corresponding API QuicStreamBufferAllocator is usually initiated in the integration code, i.e. https://cs.chromium.org/chromium/src/net/third_party/quiche/src/quic/core/quic_epoll_connection_helper.h?rcl=63d0bc4ae8dcc67502f30c320cdc074c44ea4d89&l=48

Reply:
+1 let's just keep it simple until we need something more complicated. Can we just use new/delete here for now?

Reply:
Okay, omit the passed-in allocator for now